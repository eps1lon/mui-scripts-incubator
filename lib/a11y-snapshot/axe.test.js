const playwright = require("playwright");
const { blockAds, blockSocials, gotoMuiPage } = require("./mui-utils");

/**
 * @param {import('playwright').Page} page
 */
function axeRun(page) {
	return page.evaluate(() => {
		/**
		 * @type {typeof import('axe-core')}
		 */
		const axe =
			// eslint-disable-next-line no-undef -- it's defined in the context where `playwright` runs it.
			window.axe;

		return axe.run({
			exclude: [
				// Exclude https://github.com/mui-org/material-ui/blob/aabdcaf3dcb73f2aeecb6b629e1ad675e6776ce1/docs/src/modules/components/Ad.js#L223-L230
				['[data-ga-event-category="ad"]'],
			],
		});
	});
}

describe.each(["chromium"])("%s", (browserType) => {
	/**
	 * @type {import('playwright').Browser}
	 */
	let browser;
	/**
	 * @type {import('playwright').Page}
	 */
	let page;

	beforeAll(async () => {
		browser = await playwright[browserType].launch();
	});

	afterAll(async () => {
		await browser.close();
	});

	beforeEach(async () => {
		page = await browser.newPage();
		await page.addInitScript({
			path: require.resolve("axe-core"),
		});
	});

	afterEach(async () => {
		await page.close();
	});

	it("/", async () => {
		blockSocials(page);
		await gotoMuiPage(page, "/");

		await expect(axeRun(page)).resolves.toMatchAxeSnapshot();
	});

	// markdowndocs
	it.each([
		"/api/button/",
		"/api/select/",
		"/components/breadcrumbs",
		"/components/buttons/",
		"/components/button-group/",
		"/components/checkboxes/",
		"/components/dialogs/",
		"/components/pickers",
		"/components/radio-buttons",
		"/components/selects/",
		"/components/slider",
		"/components/switches/",
		"/components/tabs/",
		"/components/text-fields/",
		"/components/tooltips/",
		"/components/transfer-list",
		// lab
		"/components/pagination/",
		"/components/rating/",
		"/components/tree-view/",
	])(`%s`, async (docsRoute) => {
		blockAds(page);
		await gotoMuiPage(page, docsRoute);

		await expect(axeRun(page)).resolves.toMatchAxeSnapshot();
	});
});
