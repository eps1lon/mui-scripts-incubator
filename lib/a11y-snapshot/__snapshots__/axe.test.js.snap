// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`chromium / 1`] = `Array []`;

exports[`chromium /api/button/ 1`] = `Array []`;

exports[`chromium /api/select/ 1`] = `Array []`;

exports[`chromium /components/breadcrumbs 1`] = `
Array [
  Object {
    "canMismatch__helpUrl": "https://dequeuniversity.com/rules/axe/4.1/landmark-unique?application=axeAPI",
    "id": "landmark-unique",
    "nodes": Array [
      "#demo-REMOVED_UNSTABLE_USE_ID_VALUE_FROM_SNAPSHOT > div[role=\\"presentation\\"] > .MuiBreadcrumbs-root.css-REMOVED_UNSTABLE_EMOTION_HASH-MuiTypography-root-MuiBreadcrumbs-root[aria-label=\\"breadcrumb\\"]",
    ],
  },
]
`;

exports[`chromium /components/button-group/ 1`] = `Array []`;

exports[`chromium /components/buttons/ 1`] = `
Array [
  Object {
    "canMismatch__helpUrl": "https://dequeuniversity.com/rules/axe/4.1/aria-progressbar-name?application=axeAPI",
    "id": "aria-progressbar-name",
    "nodes": Array [
      ".MuiLoadingButton-pendingIndicatorCenter.MuiLoadingButton-pendingIndicator > .MuiCircularProgress-root.MuiCircularProgress-indeterminate.MuiCircularProgress-colorInherit",
      ".MuiLoadingButton-pendingIndicatorStart > .MuiCircularProgress-root.MuiCircularProgress-indeterminate.MuiCircularProgress-colorInherit",
    ],
  },
  Object {
    "canMismatch__helpUrl": "https://dequeuniversity.com/rules/axe/4.1/button-name?application=axeAPI",
    "id": "button-name",
    "nodes": Array [
      ".MuiLoadingButton-pending.Mui-disabled.MuiLoadingButton-root:nth-child(1)",
      ".MuiLoadingButton-pending.Mui-disabled.MuiLoadingButton-root:nth-child(2)",
    ],
  },
]
`;

exports[`chromium /components/checkboxes/ 1`] = `
Array [
  Object {
    "canMismatch__helpUrl": "https://dequeuniversity.com/rules/axe/4.1/label?application=axeAPI",
    "id": "label",
    "nodes": Array [
      "div > .Mui-checked.css-REMOVED_UNSTABLE_EMOTION_HASH-MuiButtonBase-root-MuiIconButton-root-MuiCheckbox-root.MuiIconButton-colorSecondary > .MuiIconButton-label.css-REMOVED_UNSTABLE_EMOTION_HASH-MuiIconButton-label > input[name=\\"checkedA\\"][value=\\"\\"][data-indeterminate=\\"false\\"]",
      ".MuiIconButton-colorPrimary > .MuiIconButton-label.css-REMOVED_UNSTABLE_EMOTION_HASH-MuiIconButton-label > input[name=\\"checkedB\\"][value=\\"\\"][data-indeterminate=\\"false\\"]",
      "input[name=\\"checkedC\\"]",
      "input[name=\\"checkedD\\"]",
    ],
  },
]
`;

exports[`chromium /components/dialogs/ 1`] = `Array []`;

exports[`chromium /components/pagination/ 1`] = `
Array [
  Object {
    "canMismatch__helpUrl": "https://dequeuniversity.com/rules/axe/4.1/landmark-unique?application=axeAPI",
    "id": "landmark-unique",
    "nodes": Array [
      "#demo-REMOVED_UNSTABLE_USE_ID_VALUE_FROM_SNAPSHOT > .css-REMOVED_UNSTABLE_EMOTION_HASH > .MuiPagination-text.MuiPagination-root.css-REMOVED_UNSTABLE_EMOTION_HASH-MuiPagination-root:nth-child(1)",
    ],
  },
]
`;

exports[`chromium /components/pickers 1`] = `Array []`;

exports[`chromium /components/radio-buttons 1`] = `Array []`;

exports[`chromium /components/rating/ 1`] = `Array []`;

exports[`chromium /components/selects/ 1`] = `Array []`;

exports[`chromium /components/slider 1`] = `
Array [
  Object {
    "canMismatch__helpUrl": "https://dequeuniversity.com/rules/axe/4.1/aria-valid-attr-value?application=axeAPI",
    "id": "aria-valid-attr-value",
    "nodes": Array [
      "input[aria-labelledby=\\"discrete-slider-small-steps\\"]",
    ],
  },
  Object {
    "canMismatch__helpUrl": "https://dequeuniversity.com/rules/axe/4.1/label?application=axeAPI",
    "id": "label",
    "nodes": Array [
      "input[value=\\"\\\\31 0\\"][aria-orientation=\\"horizontal\\"][aria-valuemax=\\"\\\\31 00\\"]",
    ],
  },
]
`;

exports[`chromium /components/switches/ 1`] = `Array []`;

exports[`chromium /components/tabs/ 1`] = `Array []`;

exports[`chromium /components/text-fields/ 1`] = `Array []`;

exports[`chromium /components/tooltips/ 1`] = `Array []`;

exports[`chromium /components/transfer-list 1`] = `Array []`;

exports[`chromium /components/tree-view/ 1`] = `Array []`;
